csrc/optimization/consecutive_cast_pass.cpp

```diff
@@ -0,0 +1,157 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir/utils.h>
+#include <optimization/consecutive_cast_pass.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+bool isSameDtypeCategory(const DataType& input_t, const DataType& output_t) {
+  if ((isIntegralType(input_t) && isIntegralType(out_dtype)) ||
+      (isFloatingPointType(input_t) && isFloatingPointType(out_dtype)) ||
+      (isComplexType(input_t) && isComplexType(out_dtype))) {
+    return true;
+  }
+  return false;
+}
+
+// check if type1 is a wider type than type0
+// Which indicates a cast from type0 -> type1 -> type0 should be bit-wise identical
+bool isWiderType(const DataType& type0, const DataType& type1) {
+  if (type0 == type1) {
+    return true;
+  } else if (type0 == DataType::Double && (type1 == DataType::Float || type1 == DataType::Half || type1 == DataType::BFloat16)) {
+    return true;
+  } else if (type0 == DataType::Float && (type1 == DataType::Half || type1 == DataType::BFloat16)) {
+    return true;
+  } else if (type0 == DataType::Int && type1 == DataType::Int32) {
+    return true;
+  } else if (type0 == DataType::ComplexDouble && type1 == DataType::ComplexFloat) {
```

Comment:
I think we also have `complex double > double`, `complex float > float`, `double > int32`, `everything > bool`?